Release Android 4.0.1

Mapbox Android 4.0.1 is a follow up patch to the most ambitious Android release to date.  These new features are part of our ongoing efforts focused on providing a premiere navigation experience in-car and on-dash for Android.  

#### Highlights

* [onMapReady not getting called when offline](https://github.com/mapbox/mapbox-gl-native/issues/4500)

Release Branch: https://github.com/mapbox/mapbox-gl-native/commits/release-android-v4.0.1

#### The TODO list for making this happen:

- [x] Change Version Name to `4.0.1` in [`gradle.properties`](https://github.com/mapbox/mapbox-gl-native/blob/925687ab06892528f25fd4a79d27a55560634d96/platform/android/MapboxGLAndroidSDK/gradle.properties#L2)
- [x] Update version info in [Twitter Fabric metadata file](https://github.com/mapbox/mapbox-gl-native/blob/925687ab06892528f25fd4a79d27a55560634d96/platform/android/MapboxGLAndroidSDK/src/main/resources/fabric/com.mapbox.mapboxsdk.mapbox-android-sdk.properties#L2)
- [x] Publish `4.0.1` artifact to Maven Central
- [x] Updating Mapbox.com JavaDoc Documentation
- [x] Update [CHANGELOG.md](https://github.com/mapbox/mapbox-gl-native/blob/master/platform/android/CHANGELOG.md)
- [x] Tag `android-v4.0.1` in GitHub
- [ ] Close Milestone `android-v4.0.1` in GitHub
- [ ] ~~Blog Post on https://www.mapbox.com/blog/~~
- [ ] Tweet release
- [x] Alert Twitter Fabric team of new release
- [ ] ~~Create ticket to merge `release-androidv4.0.1` branch back into `master`~~
- [ ] ~~Change Version Name _BACK to SNAPSHOT after release and release branch merge_ to support daily development.  In this case set to  `4.1.0-SNAPSHOT` in [`gradle.properties`](https://github.com/mapbox/mapbox-gl-native/blob/925687ab06892528f25fd4a79d27a55560634d96/platform/android/MapboxGLAndroidSDK/gradle.properties#L2)~~
- [x] Update [Mapbox Demo App](https://github.com/mapbox/mapbox-android-demo) to use `4.0.1` and release on Google Play

/cc @mapbox/mobile @mapbox/support @mapbox/sales @gundersen @lxbarth  @twbell @Lndsay

I believe I've found the problem... malformed HTML in the Javadoc comments.  Originally there were 16 errors and now I after running it again after I removed 2 (from `MapboxMap`) there are 14 errors.  I'll remove the remaining 14 and we should be good to go.

```
12:27:44.935 [ERROR] [system.err] /Users/brad/development/mapbox-gl-android/platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/telemetry/MapboxEventManager.java:100: error: self-closing element not allowed
12:27:44.935 [ERROR] [system.err]      * <p/>
12:27:44.935 [ERROR] [system.err]        ^
12:27:44.956 [ERROR] [system.err] /Users/brad/development/mapbox-gl-android/platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/telemetry/TelemetryLocationReceiver.java:40: error: self-closing element not allowed
12:27:44.956 [ERROR] [system.err]      * <p/>
12:27:44.956 [ERROR] [system.err]        ^
12:27:44.966 [ERROR] [system.err] /Users/brad/development/mapbox-gl-android/platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/telemetry/TelemetryLocationReceiver.java:50: error: self-closing element not allowed
12:27:44.966 [ERROR] [system.err]      * <p/>
12:27:44.966 [ERROR] [system.err]        ^
12:27:44.990 [ERROR] [system.err] /Users/brad/development/mapbox-gl-android/platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/telemetry/TelemetryService.java:28: error: self-closing element not allowed
12:27:44.990 [ERROR] [system.err]      * <p/>
12:27:44.990 [ERROR] [system.err]        ^
12:27:44.995 [ERROR] [system.err] /Users/brad/development/mapbox-gl-android/platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/telemetry/TelemetryService.java:96: error: self-closing element not allowed
12:27:44.996 [ERROR] [system.err]      * <p/>
12:27:44.996 [ERROR] [system.err]        ^
12:27:44.996 [ERROR] [system.err] /Users/brad/development/mapbox-gl-android/platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/telemetry/TelemetryService.java:100: error: self-closing element not allowed
12:27:44.996 [ERROR] [system.err]      * <p/>
12:27:44.996 [ERROR] [system.err]        ^
12:27:44.997 [ERROR] [system.err] /Users/brad/development/mapbox-gl-android/platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/telemetry/TelemetryService.java:105: error: self-closing element not allowed
12:27:44.997 [ERROR] [system.err]      * <p/>
12:27:44.997 [ERROR] [system.err]        ^
12:27:44.997 [ERROR] [system.err] /Users/brad/development/mapbox-gl-android/platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/telemetry/TelemetryService.java:106: error: unknown tag: sample
12:27:44.997 [ERROR] [system.err]      * {@sample development/samples/ApiDemos/src/com/example/android/apis/app/ForegroundService.java
12:27:44.997 [ERROR] [system.err]        ^
12:27:44.998 [ERROR] [system.err] /Users/brad/development/mapbox-gl-android/platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/telemetry/TelemetryService.java:108: error: self-closing element not allowed
12:27:44.998 [ERROR] [system.err]      * <p/>
12:27:44.998 [ERROR] [system.err]        ^
12:27:45.330 [ERROR] [system.err] /Users/brad/development/mapbox-gl-android/platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/annotations/BaseMarkerOptions.java:9: error: self-closing element not allowed
12:27:45.330 [ERROR] [system.err]  * <p/>
12:27:45.330 [ERROR] [system.err]    ^
12:27:45.369 [ERROR] [system.err] /Users/brad/development/mapbox-gl-android/platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/annotations/Marker.java:14: error: unexpected end tag: </p>
12:27:45.369 [ERROR] [system.err]  * </p>
12:27:45.369 [ERROR] [system.err]    ^
12:27:45.370 [ERROR] [system.err] /Users/brad/development/mapbox-gl-android/platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/annotations/Marker.java:16: error: self-closing element not allowed
12:27:45.370 [ERROR] [system.err]  * <p/>
12:27:45.370 [ERROR] [system.err]    ^
12:27:45.385 [ERROR] [system.err] /Users/brad/development/mapbox-gl-android/platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/annotations/MarkerOptions.java:15: error: self-closing element not allowed
12:27:45.385 [ERROR] [system.err]  * <p/>
12:27:45.385 [ERROR] [system.err]    ^
12:27:45.385 [ERROR] [system.err] /Users/brad/development/mapbox-gl-android/platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/annotations/MarkerOptions.java:17: error: self-closing element not allowed
12:27:45.385 [ERROR] [system.err]  * <p/>
```

--- a/platform/android/CHANGELOG.md
+++ b/platform/android/CHANGELOG.md
@@ -2,6 +2,12 @@
 
 Mapbox welcomes participation and contributions from everyone.  If you'd like to do so please see the [`Contributing Guide`](https://github.com/mapbox/mapbox-gl-native/blob/master/CONTRIBUTING.md) first to get started.
 
+## 4.0.1
+
+Mapbox Android 4.0.1 is a patch release to make this bug fix available sooner.
+
+* onMapReady() not getting called when offline ([#4500](https://github.com/mapbox/mapbox-gl-native/issues/4500))
+
 ## 4.0.0
 
 Mapbox Android 4.0.0 contains the following 3 major new features.

--- a/platform/android/CHANGELOG.md
+++ b/platform/android/CHANGELOG.md
@@ -2,6 +2,14 @@
 
 Mapbox welcomes participation and contributions from everyone.  If you'd like to do so please see the [`Contributing Guide`](https://github.com/mapbox/mapbox-gl-native/blob/master/CONTRIBUTING.md) first to get started.
 
+## 4.0.0
+
+Mapbox Android 4.0.0 contains the following 3 major new features.
+
+* MapboxMap API Change ([#3145](https://github.com/mapbox/mapbox-gl-native/issues/3145))
+* Offline Maps ([#3891](https://github.com/mapbox/mapbox-gl-native/issues/3891))
+* Telemetry ([#2421](https://github.com/mapbox/mapbox-gl-native/issues/2421))
+
 ## 4.0.0-rc.1
 
 * Default Value Bug Fix for MapboxMapOptions ([#4398](https://github.com/mapbox/mapbox-gl-native/issues/4398))

--- a/platform/android/MapboxGLAndroidSDK/gradle.properties
+++ b/platform/android/MapboxGLAndroidSDK/gradle.properties
@@ -1,5 +1,5 @@
 GROUP=com.mapbox.mapboxsdk
-VERSION_NAME=4.0.0-SNAPSHOT
+VERSION_NAME=4.0.0
 
 POM_DESCRIPTION=Mapbox GL Android SDK
 POM_URL=https://github.com/mapbox/mapbox-gl-native

--- a/platform/android/MapboxGLAndroidSDK/gradle.properties
+++ b/platform/android/MapboxGLAndroidSDK/gradle.properties
@@ -1,5 +1,5 @@
 GROUP=com.mapbox.mapboxsdk
-VERSION_NAME=4.0.1-SNAPSHOT
+VERSION_NAME=4.0.1
 
 POM_DESCRIPTION=Mapbox GL Android SDK
 POM_URL=https://github.com/mapbox/mapbox-gl-native

--- a/platform/android/MapboxGLAndroidSDK/gradle.properties
+++ b/platform/android/MapboxGLAndroidSDK/gradle.properties
@@ -1,5 +1,5 @@
 GROUP=com.mapbox.mapboxsdk
-VERSION_NAME=4.0.0
+VERSION_NAME=4.0.1-SNAPSHOT
 
 POM_DESCRIPTION=Mapbox GL Android SDK
 POM_URL=https://github.com/mapbox/mapbox-gl-native

--- a/platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/annotations/BaseMarkerOptions.java
+++ b/platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/annotations/BaseMarkerOptions.java
@@ -6,7 +6,7 @@ import com.mapbox.mapboxsdk.geometry.LatLng;
 
 /**
  * Abstract builder class for composing custom Marker objects.
- * <p/>
+ *
  * Extending this class requires implementing Parceable interface.
  *
  * @param <U> Type of the marker to be composed

--- a/platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/annotations/Marker.java
+++ b/platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/annotations/Marker.java
@@ -11,9 +11,8 @@ import com.mapbox.mapboxsdk.maps.MapView;
 
 /**
  * Marker is an annotation that shows an icon image at a geographical location.
- * </p>
+ *
  * An {@link InfoWindow} can be shown when a Marker is pressed
- * <p/>
  */
 public class Marker extends Annotation {
 

--- a/platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/annotations/MarkerOptions.java
+++ b/platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/annotations/MarkerOptions.java
@@ -12,9 +12,7 @@ import com.mapbox.mapboxsdk.geometry.LatLng;
  * <p>
  * Builder for composing {@link com.mapbox.mapboxsdk.annotations.Marker} objects.
  * </p>
- * <p/>
  * <h3>Example</h3>
- * <p/>
  * <pre>
  * mMapView.addMarker(new MarkerOptions()
  *   .title("Intersection")

--- a/platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/maps/MapView.java
+++ b/platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/maps/MapView.java
@@ -400,19 +400,17 @@ public class MapView extends FrameLayout {
         addOnMapChangedListener(new OnMapChangedListener() {
             @Override
             public void onMapChanged(@MapChange int change) {
-                if (change == DID_FINISH_RENDERING_MAP_FULLY_RENDERED) {
+                if (change == DID_FINISH_RENDERING_MAP_FULLY_RENDERED && mInitialLoad) {
+                    mInitialLoad = false;
                     reloadIcons();
                     reloadMarkers();
                     adjustTopOffsetPixels();
-                    if (mInitialLoad) {
-                        mInitialLoad = false;
-                        if (mOnMapReadyCallbackList.size() > 0) {
-                            Iterator<OnMapReadyCallback> iterator = mOnMapReadyCallbackList.iterator();
-                            while (iterator.hasNext()) {
-                                OnMapReadyCallback callback = iterator.next();
-                                callback.onMapReady(mMapboxMap);
-                                iterator.remove();
-                            }
+                    if (mOnMapReadyCallbackList.size() > 0) {
+                        Iterator<OnMapReadyCallback> iterator = mOnMapReadyCallbackList.iterator();
+                        while (iterator.hasNext()) {
+                            OnMapReadyCallback callback = iterator.next();
+                            callback.onMapReady(mMapboxMap);
+                            iterator.remove();
                         }
                     }
                 }

--- a/platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/maps/MapboxMap.java
+++ b/platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/maps/MapboxMap.java
@@ -309,7 +309,7 @@ public class MapboxMap {
     /**
      * Animates the movement of the camera from the current position to the position defined in the update.
      * During the animation, a call to getCameraPosition() returns an intermediate location of the camera.
-     * <p/>
+     *
      * See CameraUpdateFactory for a set of updates.
      *
      * @param update The change that should be applied to the camera.
@@ -1136,7 +1136,7 @@ public class MapboxMap {
     /**
      * Sets the distance from the edges of the map view’s frame to the edges of the map
      * view’s logical viewport.
-     * <p/>
+     *
      * When the value of this property is equal to {0,0,0,0}, viewport
      * properties such as `centerCoordinate` assume a viewport that matches the map
      * view’s frame. Otherwise, those properties are inset, excluding part of the

--- a/platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/telemetry/MapboxEventManager.java
+++ b/platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/telemetry/MapboxEventManager.java
@@ -97,7 +97,7 @@ public class MapboxEventManager {
 
     /**
      * Internal setup of MapboxEventsManager.  It needs to be called once before @link MapboxEventManager#getMapboxEventManager
-     * <p/>
+     *
      * This allows for a cleaner getMapboxEventManager() that doesn't require context and accessToken
      *
      * @param context     The context associated with MapView

--- a/platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/telemetry/TelemetryLocationReceiver.java
+++ b/platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/telemetry/TelemetryLocationReceiver.java
@@ -37,7 +37,7 @@ public class TelemetryLocationReceiver extends BroadcastReceiver {
      * 10 seconds that the system allows before considering the receiver to
      * be blocked and a candidate to be killed). You cannot launch a popup dialog
      * in your implementation of onReceive().
-     * <p/>
+     *
      * <p><b>If this BroadcastReceiver was launched through a &lt;receiver&gt; tag,
      * then the object is no longer alive after returning from this
      * function.</b>  This means you should not perform any operations that
@@ -47,7 +47,7 @@ public class TelemetryLocationReceiver extends BroadcastReceiver {
      * {@link Context#bindService(Intent, ServiceConnection, int)}.  If you wish
      * to interact with a service that is already running, you can use
      * {@link #peekService}.
-     * <p/>
+     *
      * <p>The Intent filters used in {@link Context#registerReceiver}
      * and in application manifests are <em>not</em> guaranteed to be exclusive. They
      * are hints to the operating system about how to find suitable recipients. It is

--- a/platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/telemetry/TelemetryService.java
+++ b/platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/telemetry/TelemetryService.java
@@ -25,7 +25,7 @@ public class TelemetryService extends Service {
      * {@link IBinder} is usually for a complex interface
      * that has been <a href="{@docRoot}guide/components/aidl.html">described using
      * aidl</a>.
-     * <p/>
+     *
      * <p><em>Note that unlike other application components, calls on to the
      * IBinder interface returned here may not happen on the main thread
      * of the process</em>.  More information about the main thread can be found in
@@ -93,19 +93,17 @@ public class TelemetryService extends Service {
      * Called by the system every time a client explicitly starts the service by calling
      * {@link Context#startService}, providing the arguments it supplied and a
      * unique integer token representing the start request.  Do not call this method directly.
-     * <p/>
+     *
      * <p>For backwards compatibility, the default implementation calls
      * {@link #onStart} and returns either {@link #START_STICKY}
      * or {@link #START_STICKY_COMPATIBILITY}.
-     * <p/>
+     * </p>
      * <p>If you need your application to run on platform versions prior to API
      * level 5, you can use the following model to handle the older {@link #onStart}
      * callback in that case.  The <code>handleCommand</code> method is implemented by
      * you as appropriate:
-     * <p/>
-     * {@sample development/samples/ApiDemos/src/com/example/android/apis/app/ForegroundService.java
-     * start_compatibility}
-     * <p/>
+     * </p>
+     *
      * <p class="caution">Note that the system calls this on your
      * service's main thread.  A service's main thread is the same
      * thread where UI operations take place for Activities running in the

--- a/platform/android/MapboxGLAndroidSDK/src/main/java/overview.html
+++ b/platform/android/MapboxGLAndroidSDK/src/main/java/overview.html
@@ -1,9 +1,11 @@
 <html>
 <body>
+<p>
 An open source OpenGL-based vector map solution for Android with full styling capabilities.
-<p/>
+</p>
+<p>
 For more information, check out <a href="https://www.mapbox.com/android-sdk/">our online overview</a>.
-<p/>
+</p>
 <!--TODO: Need an Android image-->
 <!--<img src="https://raw.githubusercontent.com/mapbox/mapbox-gl-native/master/ios/screenshot.png"/>-->
 <!--<h2>Changelog</h2>

--- a/platform/android/MapboxGLAndroidSDK/src/main/resources/fabric/com.mapbox.mapboxsdk.mapbox-android-sdk.properties
+++ b/platform/android/MapboxGLAndroidSDK/src/main/resources/fabric/com.mapbox.mapboxsdk.mapbox-android-sdk.properties
@@ -1,3 +1,3 @@
 fabric-identifier=com.mapbox.mapboxsdk.mapbox-android-sdk
-fabric-version=4.0.0
+fabric-version=4.0.1
 fabric-build-type=binary

--- a/platform/android/MapboxGLAndroidSDK/src/main/resources/fabric/com.mapbox.mapboxsdk.mapbox-android-sdk.properties
+++ b/platform/android/MapboxGLAndroidSDK/src/main/resources/fabric/com.mapbox.mapboxsdk.mapbox-android-sdk.properties
@@ -1,3 +1,3 @@
 fabric-identifier=com.mapbox.mapboxsdk.mapbox-android-sdk
-fabric-version=4.0.0-rc.1
+fabric-version=4.0.0
 fabric-build-type=binary

--- a/platform/android/MapboxGLAndroidSDKTestApp/src/main/java/com/mapbox/mapboxsdk/testapp/utils/TimingLogger.java
+++ b/platform/android/MapboxGLAndroidSDKTestApp/src/main/java/com/mapbox/mapboxsdk/testapp/utils/TimingLogger.java
@@ -23,7 +23,7 @@ import android.util.Log;
 /**
  * A utility class to help log timings splits throughout a method call.
  * Typical usage is:
- * <p/>
+ *
  * <pre>
  *     TimingLogger timings = new TimingLogger(TAG, "methodA");
  *     // ... do some work A ...
@@ -34,9 +34,9 @@ import android.util.Log;
  *     timings.addSplit("work C");
  *     timings.dumpToLog();
  * </pre>
- * <p/>
+ *
  * <p>The dumpToLog call would add the following to the log:</p>
- * <p/>
+ *
  * <pre>
  *     D/TAG     ( 3459): methodA: begin
  *     D/TAG     ( 3459): methodA:      9 ms, work A

